src/internal/recovery/recovery_test.go

```diff
@@ -0,0 +1,50 @@
+package recovery
+
+import (
+	"context"
+	"strings"
+	"testing"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/clusterstate"
+	"github.com/pachyderm/pachyderm/v2/src/internal/dbutil"
+	"github.com/pachyderm/pachyderm/v2/src/internal/dockertestenv"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/chunk"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/kv"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/track"
+)
+
+func TestSnapshotDatabase(t *testing.T) {
+	ctx := pctx.TestContext(t)
+	db := dockertestenv.NewMigratedTestDB(t, clusterstate.DesiredClusterState)
+	tracker := track.NewPostgresTracker(db)
+	s := fileset.NewStorage(fileset.NewPostgresStore(db), tracker, chunk.NewStorage(kv.NewMemStore(), nil, db, tracker))
+	var snapID SnapshotID
+	if err := dbutil.WithTx(ctx, db, func(ctx context.Context, tx *pachsql.Tx) error {
+		var err error
+		snapID, err = createSnapshotRow(ctx, tx, s)
+		return errors.Wrap(err, "createSnapshotRow")
+	}); err != nil {
+		t.Fatalf("WithTx: %v", err)
```

Comment:
There are so many require in our code base. So when to use require and when to use Fatalf?

Reply:
I'd say be consistent with the existing tests in the file.  If starting a new file, avoid `require`.

Reply:
Got it.

why did we have require?

Reply:
People coming from other languages expect to write tests like that, so they added it to the codebase so they could continue doing things their own way instead of the go way.  At the end of the day it doesn't really matter, but a lot of our tests abort too soon (meaning you need to run it multiple times to gradually fix an error, rather than being able to confidently have a fixed test after reading the output) because of the `require` sugar layer.  (The open source library we "copied" has "assert" and "require" that Errorf or Fatalf depending on which one was used.  This would avoid the bailout-early test.)

The other problem is that people write things like `require.Equal(want, len(got))` which prints a message like "want: 1, got: 2" which doesn't answer the follow-up question "ok well what was in got".  Hand-written tests tend to not make this mistake.

In general you should use cmp.Diff in these cases as the diff always explains the failure as deeply as necessary to fix the test.  There is also require.NoDiff which I don't mind people using in new code, though it's a little uglier than manually calling cmp.Diff in my opinion (and often kills the test prematurely, again because require's thing is to end the test case with Fatalf on failure).

Reply:
This might be useful to be brought up in a team meeting